input-profile.ts

```diff
@@ -0,0 +1,356 @@
+
+                    if (!this._isModelLoaded()) this._loadAndMapGamepad(profile);
+                })
+                .catch((e) => {
+                    console.error(`Failed to load profile from ${this.url}. Reason:`, e);
+                });
+        });
+    }
+
+    /**
+     * Checks if the 3D controller model is loaded.
+     * @return True if the model is loaded; otherwise, false.
+     * @hidden
+     */
+    private _isModelLoaded() {
+        return this._controllerModel !== null;
+    }
+
+    /**
+     * Loads the 3D controller model and caches the mapping to the gamepad.
+     * @param profile The path to the input profile.
+     * @hidden
+     */
+    private async _loadAndMapGamepad(profile: string) {
+        const assetPath = profile + '/' + this._handedness + '.glb';
+        this._controllerModel = this.defaultController;
+        if (!this.mapToDefaultController) {
+            /** load 3d model in the runtime with profile url */
+            try {
+                this._controllerModel = (await this.engine.scene.append(
+                    assetPath
+                )) as Object3D;
+            } catch (e) {
+                console.error(
+                    `Failed to load i-p controller model. Reason:`,
+                    e,
+                    `Continuing with ${this._handedness} default controller.`
+                );
+                this._setComponentsActive(true);
+            }
+            this._controllerModel.parent = this.object;
+            this._controllerModel.setPositionLocal([0, 0, 0]);
+            this._setComponentsActive(false);
+
+            if (this.addVrModeSwitch)
+                this._controllerModel.addComponent(VrModeActiveSwitch);
+            this.onModelLoaded.notify();
+        }
+        this._cacheGamepadObjectsFromProfile(this._profileJSON, this._controllerModel);
+        this._setHandTrackingControllers(this._controllerModel);
+        this.update = () => this._mapGamepadInput();
+    }
+
+    /**
+     * Caches gamepad objects (buttons, axes) from the loaded input profile.
+     * @hidden
+     */
+    private _cacheGamepadObjectsFromProfile(profile: any, obj: Object3D) {
+        const components = profile.layouts[this._handedness].components;
+        if (!components) return;
+
+        this._buttons = [];
+        this._axes = [];
+
+        for (const i in components) {
+            const visualResponses = components[i].visualResponses;
+
+            for (const j in visualResponses) {
+                // update buttons with new interface of current visual response
+                const visualResponse = visualResponses[j];
+                const valueNode = visualResponse.valueNodeName;
+                const minNode = visualResponse.minNodeName;
+                const maxNode = visualResponse.maxNodeName;
+
+                this._gamepadObjects[valueNode] =
+                    obj.findByNameRecursive(valueNode)[0] ?? undefined;
```

Comment:
No need for `?? undefined`, since `[0]` already returns `undefined` if the array is of size `0`